SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -39,180 +39,175 @@ void SpatialEventTracer::TraceCallback(void* UserData, const Trace_Item* Item)
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.CreateDirectoryTree(*FolderPath))
 	{
-		HexStr += FString::Printf(TEXT("%02x"), SpanId.data[i]);
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Capturing trace to %s."), *FilePath);
+		Stream.Reset(Io_CreateFileStream(TCHAR_TO_ANSI(*FilePath), Io_OpenMode::IO_OPEN_MODE_WRITE));
 	}
-	return HexStr;
 }
 
-FUserSpanId SpatialEventTracer::SpanIdToUserSpanId(const Trace_SpanId& SpanId)
+SpatialEventTracer::~SpatialEventTracer()
+{
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing disabled."));
+	Trace_EventTracer_Destroy(EventTracer);
+}
+
+FUserSpanId SpatialEventTracer::GDKSpanIdToUserSpanId(const FSpatialGDKSpanId& SpanId)
 {
 	FUserSpanId UserSpanId;
-	UserSpanId.Data.SetNum(TraceSpanIdLength);
-	const int32 Size = TraceSpanIdLength * sizeof(uint8);
-	FMemory::Memcpy(UserSpanId.Data.GetData(), SpanId.data, Size);
+	UserSpanId.Data.SetNum(TRACE_SPAN_ID_SIZE_BYTES);
+	FMemory::Memcpy(UserSpanId.Data.GetData(), SpanId.GetConstId(), TRACE_SPAN_ID_SIZE_BYTES);
 	return UserSpanId;
 }
 
-TOptional<Trace_SpanId> SpatialEventTracer::UserSpanIdToSpanId(const FUserSpanId& UserSpanId)
+FSpatialGDKSpanId SpatialEventTracer::UserSpanIdToGDKSpanId(const FUserSpanId& UserSpanId)
 {
 	if (!UserSpanId.IsValid())
 	{
 		return {};
 	}
 
-	Trace_SpanId SpanId;
-	const int32 Size = TraceSpanIdLength * sizeof(uint8);
-	FMemory::Memcpy(SpanId.data, UserSpanId.Data.GetData(), Size);
-	return SpanId;
+	FSpatialGDKSpanId TraceSpanId;
+	FMemory::Memcpy(TraceSpanId.GetId(), UserSpanId.Data.GetData(), TRACE_SPAN_ID_SIZE_BYTES);
+	return TraceSpanId;
 }
 
-TOptional<Trace_SpanId> SpatialEventTracer::CreateSpan() const
+FSpatialGDKSpanId SpatialEventTracer::TraceEvent(const FSpatialTraceEvent& SpatialTraceEvent, const Trace_SpanIdType* Causes /* = nullptr*/,
+												 int32 NumCauses /* = 0*/)
 {
-	if (!IsEnabled())
+	if (Causes == nullptr && NumCauses != 0)
 	{
 		return {};
 	}
 
-	return Trace_EventTracer_AddSpan(EventTracer, nullptr, 0);
-}
+	auto MessageSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Message);
+	auto TypeSrc = StringCast<ANSICHAR>(*SpatialTraceEvent.Type.ToString());
 
-TOptional<Trace_SpanId> SpatialEventTracer::CreateSpan(const Trace_SpanId* Causes, int32 NumCauses) const
-{
-	if (!IsEnabled())
-	{
-		return {};
-	}
+	// We could add the data to this event if a custom sampling callback was used.
+	// This would allow for sampling dependent on trace event data.
+	Trace_Event Event = { nullptr, 0, MessageSrc.Get(), TypeSrc.Get(), nullptr };
 
-	if (Causes != nullptr && NumCauses > 0)
+	Trace_SamplingResult SpanSamplingResult = Trace_EventTracer_ShouldSampleSpan(EventTracer, Causes, NumCauses, &Event);
+	if (SpanSamplingResult.decision == Trace_SamplingDecision::TRACE_SHOULD_NOT_SAMPLE)
 	{
-		return Trace_EventTracer_AddSpan(EventTracer, Causes, NumCauses);
+		// Return a null SpanId so that we attempt to trace events caused by this Span.
```

Comment:
Not sure I follow this logic, could you explain?

Reply:
GetSpanId is used to get the id associated with some event. We then use that id to cause further events. In the case there is not id to get, we will return a null span id. This means the future events will be caused by a "null" spanId. TBH the comment is not needed. It was there because we shifted away from what we were doing previously. That was, if the span id could not be found, we would just not bother tracing the future caused event.